File: app/src/main/java/closer/vlllage/com/closer/handler/phone/NameHandler.java
package closer.vlllage.com.closer.handler.phone;

import java.util.Random;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.TimeAgo;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.models.GroupContact;
import closer.vlllage.com.closer.store.models.GroupInvite;
import closer.vlllage.com.closer.store.models.Phone;

public class NameHandler extends PoolMember {

    private static final String[] fallbackNames = {
        "Random Armadillo",
        "Random Aardvark",
        "Random Kangaroo",
        "Random Gorilla",
        "Random Chimpanzee",
        "Random Anaconda",
        "Random Parakeet",
        "Random Rhino",
        "Random Muskrat",
        "Random Bumblebee",
        "Random Tiger",
        "Random Ocelot",
        "Random Capybara",
        "Random Sloth",
        "Random Lemur",
        "Random Baboon"
    };

    public String getName(Phone phone) {
        if (phone == null) {
            return noName();
        }

        String name = $(Val.class).isEmpty(phone.getName()) ?
                fallbackName(phone.getId()) : phone.getName();

        if (isInactive(phone)) {
            return $(ResourcesHandler.class).getResources().getString(R.string.contact_inactive_inline, name);
        } else {
            return name;
        }
    }

    public String getName(GroupContact groupContact) {
        if (groupContact == null) {
            return noName();
        }

        String name = $(Val.class).isEmpty(groupContact.getContactName()) ?
                fallbackName(groupContact.getContactId()) : groupContact.getContactName();

        if (isInactive(groupContact)) {
            return $(ResourcesHandler.class).getResources().getString(R.string.contact_inactive_inline, name);
        } else {
            return name;
        }
    }

    public String getName(GroupInvite groupInvite) {
        if (groupInvite == null) {
            return noName();
        }

        return $(Val.class).isEmpty(groupInvite.getName()) ? noName() : groupInvite.getName();
    }

    private String fallbackName(String phoneId) {
        if (phoneId == null) {
           return noName();
        }

        int index = new Random(phoneId.hashCode()).nextInt(fallbackNames.length);
        return fallbackNames[index];
    }

    private String noName() {
        return $(ResourcesHandler.class).getResources().getString(R.string.no_name);
    }

    private boolean isInactive(GroupContact groupContact) {
        return groupContact.getContactActive().before($(TimeAgo.class).fifteenDaysAgo());
    }

    private boolean isInactive(Phone phone) {
        return phone.getUpdated().before($(TimeAgo.class).fifteenDaysAgo());
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/phone/NameHandler.kt
package closer.vlllage.com.closer.handler.phone

import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.data.DataHandler
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.GroupContact
import closer.vlllage.com.closer.store.models.GroupInvite
import closer.vlllage.com.closer.store.models.Phone
import closer.vlllage.com.closer.store.models.Phone_
import com.queatz.on.On
import io.objectbox.query.QueryBuilder
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Single
import java.util.*

class NameHandler constructor(private val on: On) {

    fun getName(phoneId: String): String {
        val phoneList = on<StoreHandler>().store.box(Phone::class).query().equal(Phone_.id, phoneId, QueryBuilder.StringOrder.CASE_SENSITIVE).build().find()
        return if (phoneList.isEmpty())
            on<ResourcesHandler>().resources.getString(R.string.unknown)
        else getName(phoneList[0])
    }

    fun getNameAsync(phoneId: String): Single<String> = on<DataHandler>().getPhone(phoneId)
                .observeOn(AndroidSchedulers.mainThread())
                .map { getName(it) }

    fun getName(phone: Phone?): String = when {
        phone == null -> noName()
        phone.name.isNullOrBlank() -> fallbackName(phone.id)
        else -> phone.name!!
    }

    fun getName(groupContact: GroupContact?): String = when {
        groupContact == null -> noName()
        groupContact.contactName.isNullOrBlank() -> fallbackName(groupContact.contactId)
        else -> groupContact.contactName!!
    }

    fun getName(groupInvite: GroupInvite?): String = if (groupInvite?.name.isNullOrBlank()) noName() else groupInvite!!.name ?: on<ResourcesHandler>().resources.getString(R.string.unknown)

    fun getFallbackName(phoneId: String? = null, name: String? = null): String = if (name.isNullOrBlank()) fallbackName(phoneId) else name

    private fun fallbackName(phoneId: String?): String {
        if (phoneId == null) {
            return noName()
        }

        val index = Random(phoneId.hashCode().toLong()).nextInt(fallbackNames.size)
        return fallbackNames[index]
    }

    private fun noName() = on<ResourcesHandler>().resources.getString(R.string.no_name)

    companion object {

        private val fallbackNames = arrayOf(
                "Random Armadillo",
                "Random Aardvark",
                "Random Kangaroo",
                "Random Gorilla",
                "Random Chimpanzee",
                "Random Anaconda",
                "Random Parakeet",
                "Random Rhino",
                "Random Muskrat",
                "Random Bumblebee",
                "Random Tiger",
                "Random Ocelot",
                "Random Capybara",
                "Random Sloth",
                "Random Lemur",
                "Random Baboon"
        )
    }
}
